### DISCUSSION TITLE
Getting custom WebRTC to work

### DISCUSSION THREAD

**Discussion by sun-path** [bluenviron/mediamtx#1581] [Questions]
I am working on a front end for a streaming application. At the moment I have an Iframe loading inside of an Anvil (python to js) framework. The Iframe loads the webrtc page served by MediaMTX. I need to get things like volume, telemetry etc about the video feed. My idea is to serve the WebRTC code directly from the Anvil server so that it can be customised to update variables on the front end (like volume etc). How would you suggest going about doing this? I understand that this is not really in the scope of the project but more broadly: when it is suggested the WebRTC page can be customised, exactly how is that accomplished? Where does the file exist? Etc. Sorry if this is a bad question. You have made a great project here.

**Comment by aler9** [comment]
<s>Hello, copy and paste the JavaScript class `Receiver` into your project:
https://github.com/aler9/mediamtx/blob/main/internal/core/webrtc_index.html

The class allows to load a WebRTC stream on a video tag of your choice, starting from a WebSocket endpoint (that in case of the server is `ws://server-ip:8889/streamname/ws`). After you do this, you can access volume by reading the tag properties:

```js
document.getElementById("video").volume
```

and you can access telemetry by using the WebRTC statistics API:

```js
this.pc.getStats(null).then((stats) => {
    console.log(stats);
});
```

A custom class is needed because there isn't a standardized way to initialize a WebRTC stream. This server uses the WebSocket way, that consists into tunneling the WebRTC handshake phase through WebSocket.

In the future, the server will support the WHIP extension, that is a standardized way to initialize a WebRTC stream, and a custom class won't be needed any more. An issue to track the development of the feature is here: https://github.com/aler9/mediamtx/issues/1429</s>

**Reply by 0xAl3xH** [reply]
Since WHIP is now supported, how can we point a `<video>` element to mediamtx now?

**Reply by aler9** [reply]
You can either copy and paste the class `WHEPReceiver` available in the [read page](https://github.com/bluenviron/mediamtx/blob/main/internal/core/webrtc_read_index.html) or use an external JS library that supports WHEP.

**Reply by legalmentemama** [reply]
How does that work? Running the webrtc_read page on its own does nothing for me, console errors out about no ICE servers being found. What should I be changing either in webrtc_read or hls_index to get this to work? :(

**Comment by krishgcek** [comment]
I had the same issue. I solved it by a reverse proxy nginx and jQuery. Basically disabled controls and used jQuery to search the element and then linked it to custom range.

**Comment by legalmentemama** [comment]
@sun-path @krishgcek I am VERY interested in any sort of guidance you could give me to access a WebRTC or HLS stream through an iframe through WS! I was trying with normal iframes but credentials cannot be embedded in any iframes so I have no idea how to proceed and my XHR requests aren't working. I didn't find any way to contact you so I would really appreciate if you could hit me up at kylerulz AT gmail or kylesiwi at Discord!

**Comment by aler9** [comment]
Informations in this discussion are out of date since the server now uses the WHIP/WHEP standard to receive/send streams through WebRTC. Updated instructions on how to use WHIP and WHEP in JavaScript are here: https://github.com/bluenviron/mediamtx/discussions/2416
